### PR TITLE
Refactors the profile route and related utils for increased clarity

### PR DESCRIPTION
**Pull request by julialucyhogan**
This PR overhauls the backend of factfinder, hopefully making the process a little clearer. 
Data fetches were split out into separate SQL calls, but in my testing none took more than 10s of millisecs, so I don't think we should see huge performance hits. Still the larger lag time is from the call to mongo.
 
might be easier to look at the new branch, as opposed to the diff, because so much changes. However, I tried to be pretty thorough in commenting. If at any point it is hard to follow, please PLEASE throw a comment in the PR. I want this to be maintainable code! :) 

related front end changes are necessary: https://github.com/NYCPlanning/labs-factfinder/pull/709

### CODE REVIEW

File: routes/profile.js [hunk 1]
```diff
@@ -1,136 +1,118 @@
...
+  if (invalidCompare(compare)) res.status(500).send({ error: 'invalid compare param' });
+
+  if (invalidProfile(profile)) res.status(500).send({ error: 'invalid profile' });
 
-router.use((req, res, next) => {
-  res.setHeader('Cache-Control', 'public, max-age=0');
-  next();
+  try {
+    const selectedGeo = await Selection.findOne({ _id });
+    const profileObj = await getProfileData(profile, selectedGeo.geoids, compare, app.db);
```
Comment (allthesignals) [inline routes/profile.js]: This is great and reads clearly... although I was a little surprised that `getProfileData` is not imported but instead defined later in this route file. I know this was prevented by a linting rule, and I have wanted to do this for smaller utility functions (things like dasherizing or do some date parsing), but `getProfileData` seems a bit more involved. I'm not quite sure what I'm getting back, and it is after all doing all the work... (maybe this is why it _should_ live here and not tucked away as a "utility"? 🦆) Thoughts?

Comment (allthesignals) [reply routes/profile.js]: Separate: I know this was how we named things before the refactor, but I'm wondering if there might be better naming to distinguish between the `profile` query parameter (provided by the route params) and the `profileObj` which is the fully generated data. Maybe we could alias it to something like `profileName`? (I realize naming is low-hanging fruit, just sorta reacting here - if it's helpful - after having not looked at any of this code in many many months :) )

File: routes/profile.js [hunk 2]
```diff
@@ -1,136 +1,118 @@
...
+ * 'change' and 'difference' calculation values.
+ * @param {string} profile - The profile type
+ * @param {Array} geoids - The list of geoids for the given selected geography
+ * @param {string} compare - Integer string representing the geoid of the comparison geography
+ * @returns {Object}
+ */
+async function getProfileData(profile, geoids, compare, db) {
+  const isAggregate = geoids.length > 1;
+
+  const queryBuilder = getQueryBuilder(profile);
```
Comment (allthesignals) [inline routes/profile.js]: See my comment above about the naming here - I'm following after a bit of confusion around when something is really a full `profile` or just the name of a profile, like part of an enum or symbol or something.

File: routes/profile.js [hunk 3]
```diff
@@ -1,136 +1,118 @@
...
+
+  // convert combined Dataframe to array of objects
+  // (change & difference calculations results in ad-hoc property addition,
+  // which is easier with objects than Dataframe rows)
+  const profileObj = profileDF.toCollection();
+
+  /* eslint-disable */
+  profileObj.map((row) => {
+    doChangeCalculations(row);
+    doDifferenceCalculations(row);
```
Comment (allthesignals) [inline routes/profile.js]: Are these difference and change _reliability_ calculations? My understanding is that `DataIngestor` handles comparison and change calcs. So these the calcs that must be _re_calculated because of the special calculations?

File: routes/profile.js [hunk 4]
```diff
@@ -1,136 +1,118 @@
...
+  });
```
Comment (allthesignals) [inline routes/profile.js]: I hope this comment isn't a duplicate buuuut can you help me understand why `map` is preferred here? What if you just returned this:

```javascript
return profileDF
  .toCollection()
  .map((row) => {
    doChangeCalculations(row);
    doDifferenceCalculations(row);
  });
```

This reads a little better to me.

File: routes/selection.js [hunk 5]
```diff
@@ -52,6 +52,7 @@ router.get('/:id', (req, res) => {
 
 
 router.post('/', (req, res) => {
+  console.log('SELECTION POST');
```
Comment (allthesignals) [inline routes/selection.js]:
```suggestion
```

File: schema/schema.sql [hunk 6]
```diff
@@ -78,7 +78,12 @@ CREATE TABLE factfinder_metadata (
 DROP TABLE IF EXISTS decennial_dictionary;
 CREATE TABLE decennial_dictionary (
   variablename text, 
-  producttype text,
+ const { CUR_YEAR, PREV_YEAR } = require('../special-calculations/data/constants');
```
Comment (allthesignals) [inline schema/schema.sql]: Awesome

File: schema/schema.sql [hunk 7]
```diff
@@ -78,7 +78,12 @@ CREATE TABLE factfinder_metadata (
...
+
+function stringifyIds(ids) {
+  if (Array.isArray(ids)) return `'${ids.join("','")}'`;
+  return `'${ids}'`;
+} producttype text,
```
Comment (allthesignals) [inline schema/schema.sql]: Actually, I'm a little confused - is this SQL or JavaScript? Not sure how this works

File: utils/data-ingestor.js [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+  /*
+   * Joins special calculation configuration to the given Dataframe,
+   * and applies aggregate calculations to the resulting Dataframe.
+   * Then, cleans up the special calculation configuration columns, and returns the final Dataframe.
+   * @param{Dataframe} d - The dataframe to operate on
+   * @returns{Dataframe}
+   */
+  recalculate(d) {
+    d = d.leftJoin(new df.DataFrame(specialCalcConfigs[this.profileType], ['variable', 'specialType']), 'variable');
```
Comment (allthesignals) [inline utils/data-ingestor.js]: Do you know what the shape of `specialCalcConfigs` is? Would it make sense to include a comment about it? I can probably just infer that it's an array of objects, pulled from the variable-specific special configurations...

File: utils/data-ingestor.js [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+    return d;
+  }
+
+  /*
+   * Turns this.data (raw SQL data) into a Dataframe, and adds columns specified by this.extraColumns
+   * @returns{Dataframe}
+   */
+  makeBaseDataFrame() {
+    let d = new df.DataFrame(this.data);
```
Comment (allthesignals) [inline utils/data-ingestor.js]: What should be the testing strategy for this? I'd love to work on some unit tests for this downstream, and I wonder how sensitive the class is to the shape of the `data`, in terms of mocking the params...

File: utils/data-ingestor.js [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+   * based on the type of 'special' row. Additionally, scales the recomputed m value
+   * if configured in 'options'
+   * @param{Row} row - The row to operate on
+   * @param{string} specialType - The type of this special variable
+   * @param{string} variable - The variable for the given row
+   * @param{string} year - The year for the given dataset
+   * @param{Object} options - Additional configuration for special calculations
+   * @returns{Row}
+   */
+  recomputeM(row, specialType, variable, year, options) {
```
Comment (allthesignals) [inline utils/data-ingestor.js]: I named a lot of variables poorly in the early iteration - should we find a better name for `recomputeM`? `recomputeSignificance`? (or is this reliability?).

File: utils/data-ingestor.js [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+   * If transformOptions exists, then applies the appropriate transformation.
+   * 'inflate' is a special transform type, which only applies to previous year data points
+   * (designated by this.isPrevious), and scales by INFLATION_FACTOR. Other allowed transformation
+   * is 'scale' type, and require transformOptions.factor; factor is applied as scalar transformation
+   * @param{Number} val - The value to transform
+   * @param{Object} [transformOptions] - Object containing 'type', and optionally 'factor'; defines the transformation
+   * @returns{Number}
+   */
+  applyTransform(val, transformOptions) {
+    if (!transformOptions) return val;
```
Comment (allthesignals) [inline utils/data-ingestor.js]: I really like this - the methods themselves contain their own validation logic, as they should. I sometimes validate on the calling side, which mixes up concerns across methods!!!

File: utils/data-ingestor.js [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+ * i.e. if special formula 'rate' is defined for value calculation 'm',
+ * the key in the formulas object would be 'm_rate'
+ */
+function getFormulaName(options, valType) {
+  if (options.formulaName) {
+    const specialFormulaName = options.formulaName[valType];
+    return `${valType}_${specialFormulaName}`;
+  }
+  return valType;
+}
```
Comment (allthesignals) [inline utils/data-ingestor.js]: Are these cols used directly in the FE? I vaguely remember your asking about how (oddly) `sum` columns were still used to contain things like "rates". Is this no longer the case?

File: utils/difference.js [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+  calculateDifferencePercents(row);
+}
+
+/*
+ * Calculate difference_sum, difference_m, and significant
+ * @param{row} - The row to do calculations for
+ */
+function calculateDifferences(row) {
+  const updatedRow = row;
+  if (exists(updatedRow.sum) && exists(updatedRow.comparison_sum)) {
```
Comment (allthesignals) [inline utils/difference.js]: "If there's a sum and a sum to compare to"

File: utils/difference.js [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+/*
+ * Calculate difference_percent, difference_percent_m, and percent_significant
+ * @param{row} - The row to do calculations for
+ */
+function calculateDifferencePercents(row) {
+  const updatedRow = row;
+  if (exists(updatedRow.percent) && exists(updatedRow.comparison_percent)) {
+    const differencePercent = executeFormula('delta', [updatedRow.percent * 100, updatedRow.comparison_percent * 100]);
+    updatedRow.difference_percent = differencePercent < 0 && differencePercent > -0.05 ? 0 : differencePercent;
```
Comment (allthesignals) [inline utils/difference.js]: Should `-0.05` be a constant?

File: utils/difference.js [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+
+    if (exists(updatedRow.percent_m) && exists(updatedRow.comparison_percent_m)) {
+      updatedRow.difference_percent_m = executeFormula('delta_m', [updatedRow.percent_m * 100, updatedRow.comparison_percent_m * 100]);
+
+      // TODO rename difference_percent_significant
+      if (updatedRow.difference_percent !== 0) updatedRow.percent_significant = executeFormula('significant', [updatedRow.difference_percent, updatedRow.difference_percent_m]);
+    }
+  }
+}
```
Comment (allthesignals) [inline utils/difference.js]: Would these logical checks be appropriate as formulas too? Or does that shift code into more unfamiliar territory?

File: utils/difference.js [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+/*
+ * Calculate difference_percent, difference_percent_m, and percent_significant
+ * @param{row} - The row to do calculations for
+ */
+function calculateDifferencePercents(row) {
+  const updatedRow = row;
+  if (exists(updatedRow.percent) && exists(updatedRow.comparison_percent)) {
+    const differencePercent = executeFormula('delta', [updatedRow.percent * 100, updatedRow.comparison_percent * 100]);
```
Comment (allthesignals) [inline utils/difference.js]: These parametric formula templates are awesome - it feels like this particular arguments should be handled in the formula itself no?

In other words, _what should be handled in JavaScript_ and _what should be handled in the Excel formula parser_? How to make that distinction?

File: utils/data-ingestor.js [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+/*
+ * The DataIngester class is used to convert profile data from postgres query into a Dataframe.
+ * Additionally for aggregate profiles, the DataIngester recalculates 'special' variables, configured
+ * in 'special-calculations/[profile].js'.
+ * NOTE: Dataframes (https://gmousse.gitbooks.io/dataframe-js/#dataframe-js) are immutable, so they cannot be modified
+ * in place. That is why all functions mutating a Dataframe also return one.
+ */
+class DataIngester {
+  /*
+   * Creates a DataIngester
```
Comment (allthesignals) [inline utils/data-ingestor.js]: nit: the file is named `data-ingestor` but the class and comment refer to `data-ingester` - could make it hard to grep things later?

File: routes/profile.js (hunk 1 above)
Comment (julialucyhogan) [reply routes/profile.js]: naming is not low hanging fruit if you ask me! I definitely think `profile` qp should become `profileName`

also, happy to extract the `getProfileData` into a separate place. I extracted it b/c i was thinking it could be repurposed in a route that does not require a selection ID but rather accepts a list of geoids -- wouldn't be used in the front end, but could be useful to people like that grad student that emailed & wanted to scrape a bunch of profile data; they could programatically hit that route, esp if we allow 'compare' to be optional. but I agree that I don't necessarily love the idea of this living in a `util`. Let's think about where else this could go ?

File: package.json [hunk 18]
```diff
@@ -15,11 +15,13 @@
   "license": "ISC",
   "dependencies": {
     "apicache": "^1.2.3",
+    "babel-cli": "^6.26.0",
```
[the remark this answers is not in the document]
Comment (julialucyhogan) [reply package.json]: i think this was an accident :) will clean up

File: routes/profile.js (hunk 3 above)
Comment (julialucyhogan) [reply routes/profile.js]: No sorry, i should clarify. The `DataIngestor` handles all `sum` and `m` recalculations for aggregate datasets, acting on each dataset (current, `previous`, `comparison`) separately. Then, all change & difference values are calculated once the three datasets are joined. This includes everything from `change_sum`, `difference_sum` etc down to `percent_significant` and `percentage_point_significant` -- these utils do it all. I can add some comments to make that more clear. 

I'm also now realizing that I don't recalculate cv in my branch!!! need to add that !

File: routes/profile.js (hunk 4 above)
Comment (julialucyhogan) [reply routes/profile.js]: agree

File: schema/schema.sql (hunk 7 above)
Comment (julialucyhogan) [reply schema/schema.sql]: this is javascript -- pretty much exactly your stringify function, but i added handling for if `ids` was one value instead of an array, because I'm using the same query for single and aggregate profile selections. 

I think that it might be better tho to format the whole where clause for geoids in a function, because I believe SQL engines can perform much better given `column = VALUE` as opposed to `column IN (VALUE)` , and why not help the SQL engine out?

File: utils/data-ingestor.js [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+   * Turns this.data (raw SQL data) into a Dataframe, and adds columns specified by this.extraColumns
+   * @returns{Dataframe}
+   */
+  makeBaseDataFrame() {
+    let d = new df.DataFrame(this.data);
+    this.extraColumns.forEach((col) => {
+      d = d.withColumn(col);
+    });
+    return d;
+  }
```
[the remark this answers is not in the document]
Comment (julialucyhogan) [reply utils/data-ingestor.js]: It happened this way mostly b/c I got around to dealing with renaming the columns after I had created the class constructor. I don't really envision this class ever being instantiated and then reused; it is mostly a class so that I could cleanly & easily access the class properties across functions, without always having to pass them down. 

upon reflection, i agree -- the column prefix string should be part of the construction b/c why not!

File: utils/data-ingestor.js (hunk 8 above)
Comment (julialucyhogan) [reply utils/data-ingestor.js]: will comment !

File: utils/data-ingestor.js (hunk 9 above)
Comment (julialucyhogan) [reply utils/data-ingestor.js]: as long as data is an array of objects with the same named properties, a dataframe will be created. However, I will add a comment about what the data is expected to look like and that should make testing easier. I definitely DEFINITELY want your help there :)

File: utils/data-ingestor.js (hunk 10 above)
Comment (julialucyhogan) [reply utils/data-ingestor.js]: recomputeM is a function that recomputes the variable named 'm'... the variable named 'm' should be renamed, and that should be much easier now at least in the backend, but I thought that could wait for a separate PR

File: utils/difference.js (hunk 14 above)
Comment (julialucyhogan) [reply utils/difference.js]: yes !

File: utils/difference.js (hunk 15 above)
Comment (julialucyhogan) [reply utils/difference.js]: these conditions can be added to the formulas if we think that's the correct implementation! 

I put them here because it saves us some computation -- we exit early & don't do any formula parsing at all if the values are not set. However, if it is clearer to have the boolean checks in the execl formulas, I'd say that's more important. I doubt this 'optimization' has any affect on the bottom line timing anyway.

File: utils/difference.js (hunk 16 above)
Comment (julialucyhogan) [reply utils/difference.js]: I did this check outside of the formula in order to be able to reuse the same 'delta' formula that I use in all the other calculations; again if we think it's clearer to do it in a formula it would be very easy to add another formula to this list and use that here !

File: utils/data-ingestor.js (hunk 17 above)
Comment (julialucyhogan) [reply utils/data-ingestor.js]: def, will fix the typo

File: utils/data-ingestor.js (hunk 12 above)
Comment (julialucyhogan) [reply utils/data-ingestor.js]: sorry this is a confusing one... what we're returning here is the name of the formula to parse... the executeFormula() function accepts a formula NAME as a param, not the actual formula itself. This function does some logic to handle the fact that a few variables don't use the straight up `sum` or `m` formulas, but rather `sum_[specialFormulaName]` or `m_[specialFormulaName]`. This function attempts to obfuscate that tricky implementation detail when you're reading thru the higher level functions, since it is just that -- an implementation detail. I can try to make the comment more clear.

File: schema/schema.sql (hunk 7 above)
Comment (allthesignals) [reply schema/schema.sql]: @julialucyhogan I guess I'm still confused because I don't know how this would parse:
https://github.com/NYCPlanning/labs-factfinder-api/blob/7d0b125ce186ea335d0544dfb3f91a976f3bfe59/schema/schema.sql#L79-L93

It's like a random JS function was pasted into the SQL query D: sorry if I'm still missing something.

File: utils/data-ingestor.js (hunk 9 above)
Comment (allthesignals) [reply utils/data-ingestor.js]: Cool! I can start a branch somewhere for testing :D

File: routes/profile.js [hunk 20]
```diff
@@ -32,52 +32,64 @@ router.get('/:id/:profile', async (req, res) => {
...
   profileDF = profileDF.join(
     // 'compare' profiles are always NOT aggregate; they are comprised of a single geoid
-    new DataIngestor(compareProfileData, profile).processRaw('comparison'),
+    new DataIngestor(
+      compareProfileData,
+      profileName,
+      /* is aggregate */ false,
+      /* is previous */ false,
+      /* is comparison */ true,
+    ).processRaw(),
```
Comment (allthesignals) [inline routes/profile.js]: One API enhancement we could do is change the parameters of `DataIngestor` to expect an object for _optional_ params like is aggregate, previous, comparision, etc:

```javascript
new DataIngestor(compareProfileData, profileName, {
  isAggregate: false,
  isPrevious: false,
  isComparison: true,
}).processRaw(),
```

File: routes/profile.js [hunk 21]
```diff
@@ -32,52 +32,64 @@ router.get('/:id/:profile', async (req, res) => {
...
-  /* eslint-enable */
-
-  return profileObj;
+  return profileDF
+    .toCollection()
+    .map((row) => {
+      doChangeCalculations(row);
+      doDifferenceCalculations(row);
+      return row;
+    });
```
Comment (allthesignals) [inline routes/profile.js]: Great!!
